### PR TITLE
[Kubernetes][Container Logs] Add ability to configure `data_stream.dataset` & "Custom configuration"

### LINKED ISSUE #5551: Allow customizing the `data_stream.dataset` of Kubernetes container logs
Today, it seems to be [recommended](https://github.com/elastic/integrations/issues/4466#issuecomment-1403173070) to use the [Kubernetes Container Logs integration](https://github.com/elastic/integrations/tree/main/packages/kubernetes/data_stream/container_logs) for collecting container logs. One of this issues with this integration is that you currently can't specify a `data_stream.dataset` value, like you can with something like the [Log integration](https://github.com/elastic/integrations/tree/main/packages/log/data_stream/log). I'd like to propose adding the ability to configure this option, so that it is easier to use the integration to collect and process custom logs from containers.

### PR DESCRIPTION
**Pull request by BenB196**
Enhancement

## What does this PR do?

This PR does 2 things for the Kubernetes Container Log integration:

1. Adds the ability to configure `data_stream.dataset`.
2. Adds the ability to add custom configurations to container logs

Both changes are intended to bring this integration closer inline with the [Custom Log](https://github.com/elastic/integrations/tree/main/packages/log) integration.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- ~~[ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).~~

## Author's Checklist

- `data_stream.dataset` is now configurable
- "Custom configuration" section injects additional yaml configuration.

<details>
<summary>Before Upgrade</summary>

```yaml
inputs:
  - id: filestream-container-logs-329e9d5f-dd99-438e-8271-3e3598c8300a
    name: kubernetes-1
    revision: 1
    type: filestream
    use_output: default
    meta:
      package:
        name: kubernetes
        version: 1.32.0
    data_stream:
      namespace: default
    package_policy_id: 329e9d5f-dd99-438e-8271-3e3598c8300a
    streams:
      - id: >-
          kubernetes-container-logs-${kubernetes.pod.name}-${kubernetes.container.id}
        data_stream:
          dataset: kubernetes.container_logs
          type: logs
        paths:
          - '/var/log/containers/*${kubernetes.container.id}.log'
        prospector.scanner.symlinks: true
        parsers:
          - container:
              stream: all
              format: auto
```

</details>

<details>
<summary>After Upgrade (No modifications)</summary>

```yaml
inputs:
  - id: filestream-container-logs-329e9d5f-dd99-438e-8271-3e3598c8300a
    name: kubernetes-1
    revision: 2
    type: filestream
    use_output: default
    meta:
      package:
        name: kubernetes
        version: 1.33.0
    data_stream:
      namespace: default
    package_policy_id: 329e9d5f-dd99-438e-8271-3e3598c8300a
    streams:
      - id: >-
          kubernetes-container-logs-${kubernetes.pod.name}-${kubernetes.container.id}
        data_stream:
          dataset: kubernetes.container_logs
        prospector.scanner.symlinks: true
        paths:
          - '/var/log/containers/*${kubernetes.container.id}.log'
        parsers:
          - container:
              stream: all
              format: auto
```

</details>

**Note:** After the upgrade, the policy configuration no longer is setting `streams.data_stream.type: logs`. I don't believe this matters as I check the Custom logs integration and it does the same thing, as well as the static tests passed. But if desired, I can have it add `streams.data_stream.type: logs`.

<details>
<summary>After Upgrade (With modifications)</summary>

```yaml
inputs:
  - id: filestream-container-logs-329e9d5f-dd99-438e-8271-3e3598c8300a
    name: kubernetes-1
    revision: 4
    type: filestream
    use_output: default
    meta:
      package:
        name: kubernetes
        version: 1.33.0
    data_stream:
      namespace: default
    package_policy_id: 329e9d5f-dd99-438e-8271-3e3598c8300a
    streams:
      - id: >-
          kubernetes-container-logs-${kubernetes.pod.name}-${kubernetes.container.id}
        data_stream:
          dataset: custom.dataset
        prospector.scanner.symlinks: true
        paths:
          - '/var/log/containers/*${kubernetes.container.id}.log'
        file_identity.inode_marker.path: /logs/.filebeat-marker #<-- Custom configuration line here.
        parsers:
          - container:
              stream: all
              format: auto
```

</details>

## How to test this PR locally

1. Install current integration
2. Upgrade integration to this PR
3. Observe that policy is similar to before the upgrade and the integration works
4. Change `data_stream.dataset` value, and observe change is reflected in agent policy
5. Change Custom configuration value, and observe change is reflected in agent policy

## Related issues

- Closes #5551

## Screenshots

image

### CODE REVIEW

File: packages/kubernetes/data_stream/container_logs/agent/stream/stream.yml.hbs [hunk 1]
```diff
@@ -16,4 +18,6 @@ parsers:
 {{#if processors}}
 processors:
 {{processors}}
-{{/if}}
\ No newline at end of file
+{{/if}}
+
+{{custom}}
```
Comment (ChrsMark) [inline packages/kubernetes/data_stream/container_logs/agent/stream/stream.yml.hbs:23]: How about surrounding this with an `{{if}}` block?

File: packages/kubernetes/changelog.yml [hunk 2]
```diff
@@ -1,4 +1,12 @@
 # newer versions go on top
+- version: "1.33.0"
+  changes:
+    - description: Add ability to configure `data_stream.dataset`.
+      type: enhancement
+      link: https://github.com/elastic/integrations/pull/5555
+    - description: Add ability to configure additional settings via "Custom configuration" section.
```
Comment (ChrsMark) [inline packages/kubernetes/changelog.yml:7]: Maybe combine this into one message? Sth like `Add data_stream.dataset setting and custom yaml input`.

@gizas do you maybe know if we prefer spitted messages?

**Review by ChrsMark — COMMENTED** [review]
These additions look good to me. 
@gizas @tommyers-elastic raising this with you too in case you have any additional comment or objection.

**Comment by gizas** [comment]
> These additions look good to me. @gizas @tommyers-elastic raising this with you too in case you have any additional comment or objection.

As long as we introduce the dataset change, I would only advise to add the same functionality to all rest datastreams. It would be strange not to have the same experience to all.

File: packages/kubernetes/data_stream/container_logs/manifest.yml [hunk 3]
```diff
@@ -19,6 +19,13 @@ streams:
         required: true
         show_user: true
         default: true
+      - name: data_stream.dataset
+        required: true
+        default: kubernetes.container_logs
+        title: Dataset name
+        description: >
+          Set the name for your dataset. Changing the dataset will send the data to a different index. You can't use `-` in the name of a dataset and only valid characters for [Elasticsearch index names](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-index_.html).
```
Comment (gsantoro) [inline packages/kubernetes/data_stream/container_logs/manifest.yml:27]:
```suggestion
          Set the name for your dataset. Changing the dataset will send the data to a different index. For more info look at [data_stream field](https://www.elastic.co/guide/en/ecs/master/ecs-data_stream.html).
```

File: packages/kubernetes/data_stream/container_logs/agent/stream/stream.yml.hbs (hunk 1 above)
Comment (gsantoro) [reply packages/kubernetes/data_stream/container_logs/agent/stream/stream.yml.hbs:23]: For reference the [fleet_server integration](https://github.com/elastic/integrations/blob/58b500ca847590f1d33fa399cb03f023a2026490/packages/fleet_server/agent/input/agent.yml.hbs#L15) uses

```yaml
{{#if custom}}
{{custom}}
{{/if}}
```

Comment (gsantoro) [reply packages/kubernetes/data_stream/container_logs/agent/stream/stream.yml.hbs:23]: Like @gizas commented in the PR, if you are doing this, do you mind doing the same for every datastream under the kubernetes package?

**Review by gsantoro — CHANGES_REQUESTED** [review]

**Comment by BenB196** [comment]
>

**Comment by gizas** [comment]
We had another internal round of discussions and I would like to remove my comment of https://github.com/elastic/integrations/pull/5555#issuecomment-1477421219

The scenario that the user might only retrieve logs from specific application/ pods etc and would like to create a specific dataset to only include eg nginx logs is valid. But lets not expose the change to metrics, as those will still be coming from kubernetes and would create more confusion.

Sorry for the mess here.

So also this change https://github.com/elastic/integrations/pull/5555#discussion_r1143152040, to be done only for logs

**Review by gsantoro — APPROVED** [review]

**Review by ChrsMark — APPROVED** [review]
lgtm, thanks!

**Review by gizas — APPROVED** [review]

**Pull request merged**